### ISSUE TITLE
Gait Pattern

### ISSUE THREAD

**Issue by sukhijab** [cmm-20/final-project#12]
Hi,

Our group is kind of stuck in the gait pattern implementation.
So what we have so far is that we have an inverse kinematics object. It takes in the target points for all end effectors and calculates via newton’s method the joint angles. 
This seems to be working fine.

For the gaits, we use tripod. Where the y position is a cosine curve and the x/z position is a constant step. Thus, over half a period of a cosine (where we are back on the floor again) 3 pair of legs perform the forward movement (using the swing in the y direction as defined).
Then we repeat this for the other 3 pair of legs.

This seems to fail drastically. First of all the collision of the leg on the floor (y=0) is very impulsive. However, taking small amplitudes does help in reducing the impulsive effect but nonetheless, together with the forward movement this just doesnt work.

We would highly appreciate your feedback/help on our approach.

Thanks a lot

**Comment by nathanael-k** [comment]
Hey, if you are still struggling, maybe I can help with some ideas:

- are you basing your calculations for the end of the legs in local coordinates or in global coordinates? In my project I have some internal coordinate system for the legs. Depending on your setup, if you for example ever set a global coordinate slightly below 0 in y direction, then the motors would try to push the legs below the ground, which will fail. So this is where I would start: Try to find exactly when it fails, you can also time step trough the simulation (note that rendering frame rate is 30, while physics frame rate is 120).

But your basic approach should work, there is nothing wrong with it.

**Comment by Yingyan-Xu** [comment]
Do you use finite difference to get jacobian / hessian or analytical solutions? We are using FD but amazingly slow...

**Comment by nathanael-k** [comment]
Personally, I use an analytical / geometrical solution for the IK. That is very fast as is expected. Prof. Coros drops a few hints in the video.
But i think there is no reason why a similar approach as in one of the first assignment would be too slow for 6 legs. Are you maybe iterating the optimization only once per app frame? Because that would be too slow. Be sure to iterate many times per frame that gets rendered.

**Comment by sukhijab** [comment]
Thanks for the tips guys. Luckily, I was able to figure out what was going on and well now the thing moves as desired. However, it is no way close to perfect. There are still some things I am trying out. Maybe I ll post some more doubts if I have any :)
I used the analytical jacobian, which is luckily already provided to us. However, even with finite differences the simulation is running fairly well (I am running it on an i9 8 core processor though tbh).
